editor/editor_command_palette.cpp

```diff
@@ -0,0 +1,246 @@
+	}
+	Command p_command;
+	p_command.name = p_command_name;
+	p_command.callable = p_action.bind(argptrs, arguments.size());
+
+	commands[p_key_name] = p_command;
+}
+
+void EditorCommandPalette::execute_command(String p_command_name) {
+	ERR_FAIL_COND_MSG(!commands.has(p_command_name), p_command_name + " not found.");
+	commands[p_command_name].callable.call_deferred(nullptr, 0);
+}
+
+void EditorCommandPalette::register_shortcuts_as_command() {
+	const String *p_key = nullptr;
+	p_key = unregisterd_shortcuts.next(p_key);
+	while (p_key != nullptr) {
+		String command_name = unregisterd_shortcuts[*p_key].first;
+		Ref<Shortcut> p_shortcut = unregisterd_shortcuts[*p_key].second;
+		Ref<InputEventShortcut> ev;
+		ev.instantiate();
+		ev->set_shortcut(p_shortcut);
+		add_command(command_name, *p_key, callable_mp(EditorNode::get_singleton()->get_viewport(), &Viewport::unhandled_input), varray(ev, false));
+		p_key = unregisterd_shortcuts.next(p_key);
+	}
+	unregisterd_shortcuts.clear();
+}
+
+Ref<Shortcut> EditorCommandPalette::add_shortcut_command(const String &p_command, const String &p_key, Ref<Shortcut> p_shortcut) {
+	if (is_inside_tree()) {
+		Ref<InputEventShortcut> ev;
+		ev.instantiate();
+		ev->set_shortcut(p_shortcut);
+		add_command(p_command, p_key, callable_mp(EditorNode::get_singleton()->get_viewport(), &Viewport::unhandled_input), varray(ev, false));
+	} else {
+		const String command_name = String(p_command);
+		const String key_name = String(p_key);
+		Pair p_pair = Pair(command_name, p_shortcut);
+		unregisterd_shortcuts[key_name] = p_pair;
+	}
+	return p_shortcut;
+}
+
+EditorCommandPalette *EditorCommandPalette::get_singleton() {
+	if (singleton == nullptr) {
+		singleton = memnew(EditorCommandPalette);
+	}
+	return singleton;
+}
+
+EditorCommandPalette::EditorCommandPalette() {
+	allow_multi_select = false;
+
+	VBoxContainer *vbc = memnew(VBoxContainer);
+	// todo: add theme change signal handler
+	add_child(vbc);
+
+	command_search_box = memnew(LineEdit);
+	command_search_box->set_placeholder("search for a command");
+	command_search_box->set_placeholder_alpha(0.5);
+	command_search_box->connect("gui_input", callable_mp(this, &EditorCommandPalette::_sbox_input));
+	command_search_box->connect("text_changed", callable_mp(this, &EditorCommandPalette::_update_command_search).unbind(1));
+	command_search_box->connect("text_submitted", callable_mp(this, &EditorCommandPalette::_confirmed).unbind(1));
+	vbc->add_margin_child(TTR("Search:"), command_search_box);
+	register_text_enter(command_search_box);
+
+	search_options = memnew(Tree);
+	search_options->connect("item_activated", callable_mp(this, &EditorCommandPalette::_confirmed));
+	search_options->create_item();
+	search_options->set_hide_root(true);
+	search_options->set_hide_folding(true);
+	search_options->add_theme_constant_override("draw_guides", 1);
+	vbc->add_margin_child(TTR("Matches:"), search_options, true);
+
+	command_keys = memnew(List<String>);
```

Comment:
This still needs to be freed, it's a pointer to a list. You still need to call `memdelete(...)` to free it.
You should define a C++ destructor where to call it: 
```
EditorCommandPalette::~EditorCommandPalette() {
  memdelete(command_keys);
}
```